### PR TITLE
Procedural materials, procedurals on models, v3 procedural shaders, procedural vertex shaders, shader fixes, migrate shape procedurals from userData to proceduralData

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.manuscript.com/f/cases/7120/Ability-for-a-user-to-create-import-custom-shaders-as-an-entity-material-type
https://highfidelity.manuscript.com/f/cases/1435/entities-with-procedural-shaders-should-use-true-first-class-property
https://highfidelity.fogbugz.com/f/cases/14963/Add-support-for-shaders-on-meshes-and-avatars

Test plan:
# V3 Shaders
- Perform these tests on various graphics cards: NVIDIA, AMD, test on Mac
- Run this from your console:
```
Entities.addEntity({
	type: "Box",
	position: Vec3.sum(MyAvatar.position, Vec3.multiply(4.0, Quat.getForward(MyAvatar.orientation))),
	dimensions: {x: 0.5, y: 0.5, z: 0.5},
	collisionless: true,
	proceduralData: JSON.stringify({"version":3,"shaderUrl":"https://gist.githubusercontent.com/SamGondelman/932c8c03fd39b0549f0a07c0a736abd9/raw/becf86c3b1c83125c785125a781d2013a1c7c3f3/raymarch.fs","uniforms":{"scene":0}})
})
```
- You should see some grey spheres moving around a green rippling ball that morphs into a beach ball.  They should be lit correctly by the keylight (try changing the keylight.  The shadowed areas on the spheres should change).  Try placing other lights and entities around.  The spheres should be lit by the local lights and should depth-test properly with the objects.
- Look at the balls in VR.  They should appear 3D.  Pass your hand through them.  They should intersect your hand properly (with respect to depth, not physics).
# Shape/skybox procedurals
- Run the procedural entity test from this hifi_tests PR: https://github.com/highfidelity/hifi_tests/tree/6a1b4a94ac4fc8873aec66da6db670ed4b3a9875/tests/content/entity/procedural
- Run [this](https://gist.githubusercontent.com/SamGondelman/d2fe968f275be983ba2ed328437f1add/raw/692108224ec0ab4cfc5b2fd4f74fe42e8411624d/CreateProcedurals.js).

TODO ADD PICTURES.

# Procedural Materials on shapes/models/avatars
- Run [this](https://gist.githubusercontent.com/SamGondelman/8a76e0c112201526b9f21bbd33c1e1d5/raw/fef86db3b898ccae727d3e9b797f189a95d469c6/ProceduralMaterials.js).

TODO ADD PICTURES

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12149/HighFidelity-Beta-PR13823-834c29a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12149/HighFidelity-Beta-Interface-PR13823-834c29a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12149/HighFidelity-Beta-PR13823-834c29a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12149/HighFidelity-Beta-Interface-PR13823-834c29a.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12161/HighFidelity-Beta-PR13823-b95e38c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12161/HighFidelity-Beta-Interface-PR13823-b95e38c.exe

File: libraries/entities/src/EntityTree.cpp [hunk 1]
```diff
@@ -2589,6 +2589,23 @@ bool EntityTree::readFromMap(QVariantMap& map) {
...
+            QJsonObject proceduralData;
+
+            static const QString PROCEDURAL_USER_DATA_KEY = "ProceduralEntity";
+            QJsonValue proceduralEntity = userData[PROCEDURAL_USER_DATA_KEY];
+            if (!proceduralEntity.isNull()) {
+                proceduralData = proceduralEntity.toObject();
+                userData.remove(PROCEDURAL_USER_DATA_KEY);
+            }
+
+            properties.setProceduralData(QJsonDocument(proceduralData).toJson());
```
Comment (jherico) [inline libraries/entities/src/EntityTree.cpp]: what happens if the user enters a domain with old style procedural data and doesn't have edit rights?

**Comment by jherico** [comment]
why are there so many new header inter-dependencies?

File: libraries/entities/src/EntityTree.cpp (hunk 1 above)
Comment (SamGondelman) [reply libraries/entities/src/EntityTree.cpp]: This is done on the entity server, when it reads old data from a JSON file

**Comment by AndrewMeadows** [comment]
This PR needs more labels.  Surely there are some relevant labels that have not yet been applied.  Go for the record!

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12179/HighFidelity-Beta-PR13823-81345be.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12179/HighFidelity-Beta-Interface-PR13823-81345be.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12179/HighFidelity-Beta-PR13823-81345be.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12179/HighFidelity-Beta-Interface-PR13823-81345be.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12183/HighFidelity-Beta-PR13823-ceefa28.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12183/HighFidelity-Beta-Interface-PR13823-ceefa28.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12184/HighFidelity-Beta-PR13823-12e2ac0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12184/HighFidelity-Beta-Interface-PR13823-12e2ac0.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12186/HighFidelity-Beta-PR13823-4ca63ef.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12186/HighFidelity-Beta-Interface-PR13823-4ca63ef.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12186/HighFidelity-Beta-PR13823-4ca63ef.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12186/HighFidelity-Beta-Interface-PR13823-4ca63ef.exe

**Comment by vegaslon** [comment]
Just want to note. Running this PR on a windows 10 machine, at this time, with a AMD graphics card causes the transparent on the teleport textures in the default serverless domain to not show correctly.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12285/HighFidelity-Beta-PR13823-42ee601.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12285/HighFidelity-Beta-Interface-PR13823-42ee601.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12285/HighFidelity-Beta-PR13823-42ee601.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12285/HighFidelity-Beta-Interface-PR13823-42ee601.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12304/HighFidelity-Beta-PR13823-d01655a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12304/HighFidelity-Beta-Interface-PR13823-d01655a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12304/HighFidelity-Beta-PR13823-d01655a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12304/HighFidelity-Beta-Interface-PR13823-d01655a.exe

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13741/HighFidelity-Beta-PR13823-1234b1f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13741/HighFidelity-Beta-Interface-PR13823-1234b1f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13759/HighFidelity-Beta-PR13823-a1da120.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13759/HighFidelity-Beta-Interface-PR13823-a1da120.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13759/HighFidelity-Beta-PR13823-a1da120.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13759/HighFidelity-Beta-Interface-PR13823-a1da120.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13810/HighFidelity-Beta-PR13823-aa437d1.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13815/HighFidelity-Beta-PR13823-2749eed.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13815/HighFidelity-Beta-PR13823-2749eed.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13815/HighFidelity-Beta-Interface-PR13823-2749eed.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14444/HighFidelity-Beta-PR13823-fb4d774.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14444/HighFidelity-Beta-Interface-PR13823-fb4d774.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14479/HighFidelity-Beta-PR13823-e51f32e.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14480/HighFidelity-Beta-PR13823-09445af.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14480/HighFidelity-Beta-PR13823-09445af.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14480/HighFidelity-Beta-Interface-PR13823-09445af.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14480/HighFidelity-Beta-PR13823-09445af.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14480/HighFidelity-Beta-Interface-PR13823-09445af.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15024/HighFidelity-Beta-PR13823-58cdf9b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15024/HighFidelity-Beta-PR13823-58cdf9b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15024/HighFidelity-Beta-Interface-PR13823-58cdf9b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15024/HighFidelity-Beta-PR13823-58cdf9b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15024/HighFidelity-Beta-Interface-PR13823-58cdf9b.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15025/HighFidelity-Beta-PR13823-45c05b0.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15025/HighFidelity-Beta-PR13823-45c05b0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15025/HighFidelity-Beta-Interface-PR13823-45c05b0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15025/HighFidelity-Beta-PR13823-45c05b0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15025/HighFidelity-Beta-Interface-PR13823-45c05b0.exe

**Comment by hifi-gustavo** [comment]
Can someone from High Fidelity approve this PR?

**Pull request closed without merging**